### PR TITLE
msftidy_docs

### PR DESCRIPTION
**Pull request by h00die**
This PR adds a new tool: `msftidy_docs.rb`
It acts like `msftidy` but is for docs.
Part of the greater cleanup: #12853 and partially implements https://github.com/rapid7/metasploit-framework/pull/12831#issuecomment-577292896

Right now it checks for the following things:

1. Checks to make sure the doc has a corresponding rb/py/go module
2. checks ## Vulnerable Application is the first line
3. checks to make sure Vulnerable Application, Verification Steps, Scenarios, Options are all present.
4. Checks if Description or Intro/Introduction are present and recommends moving them under Vulnerable Application
5. Checks newline at end of file
6. Checks the H2s mentioned in 3 are in the right order (@tperry-r7 should it be OK to be missing a section, or all all required? Right now i have all required)
7. checks for spaces EOL (when not in ``` or 4x[:space:])
8. Checks that there are no H1s
9. Checks line length and suggests cutting it at 140 (arbitrary number I picked, up for debate)

### CODE REVIEW

File: tools/dev/msftidy_docs.rb [hunk 1]
```diff
@@ -0,0 +1,287 @@
+#!/usr/bin/env ruby
```
Comment (wvu-r7) [inline tools/dev/msftidy_docs.rb]: This file should be `+x`.

**Review by wvu — CHANGES_REQUESTED** [review]

File: tools/dev/msftidy_docs.rb [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+        warn("Spaces at EOL", idx)
+      end
+
+      if ln =~ /^# /
+        warn("No H1 (#) headers.  If this is code, indent.", idx)
+      end
+
+      l = 140
+      if ln.length > l
+        warn("Line too long (#{ln.length}).  Consider a newline (which resolves to a space in markdown) to break it up around #{l} characters.", idx)
```
Comment (wvu-r7) [inline tools/dev/msftidy_docs.rb]: FYI, this doesn't take into account code blocks.

**Comment by wvu-r7** [comment]
```
wvu@kharak:/rapid7/metasploit-framework:feature/wordpress$ tools/dev/msftidy_docs.rb documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md
documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md - [WARNING] Missing Section: ## Verification Steps
documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md - [WARNING] Descriptions should be within Vulnerable Application, or an H3 sub-section of Vulnerable Application
documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md - [WARNING] Intro/Introduction should be within Vulnerable Application, or an H3 sub-section of Vulnerable Application
documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md - [WARNING] H2 headings in incorrect order. Should be: Vulnerable Application, Verification Steps, Options, Scenarios
documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md:57 - [WARNING] Line too long (1096). Consider a newline (which resolves to a space in markdown) to break it up around 140 characters.
wvu@kharak:/rapid7/metasploit-framework:feature/wordpress$
```

https://github.com/rapid7/metasploit-framework/pull/12853#discussion_r369843771

**Comment by wvu-r7** [comment]
```
wvu@kharak:/rapid7/metasploit-framework:feature/wordpress$ tools/dev/msftidy_docs.rb documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md
documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md - [WARNING] Missing Section: ## Verification Steps
documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md - [WARNING] Descriptions should be within Vulnerable Application, or an H3 sub-section of Vulnerable Application
documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md - [WARNING] Intro/Introduction should be within Vulnerable Application, or an H3 sub-section of Vulnerable Application
documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md - [WARNING] H2 headings in incorrect order. Should be: Vulnerable Application, Verification Steps, Options, Scenarios
documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md:59 - [WARNING] Line too long (1096). Consider a newline (which resolves to a space in markdown) to break it up around 140 characters.
wvu@kharak:/rapid7/metasploit-framework:feature/wordpress$ git diff
diff --git a/documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md b/documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md
index 7857c3bfe6..8e3143ffdd 100644
--- a/documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md
+++ b/documentation/modules/exploit/unix/webapp/wp_infinitewp_auth_bypass.md
@@ -1,5 +1,7 @@
 ## Vulnerable Application

+### Description
+
 This module exploits an authentication bypass in the WordPress
 InfiniteWP Client plugin to log in as an administrator and execute
 arbitrary PHP code by overwriting the file specified by `PLUGIN_FILE`.
wvu@kharak:/rapid7/metasploit-framework:feature/wordpress$
```

File: tools/dev/msftidy_docs.rb [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+    unless has_scenarios
+      warn('Missing Section: ## Scenarios')
+    end
+
+    unless has_options
+      warn('Missing Section: ## Options')
+    end
+
+    unless has_bad_description
```
Comment (wvu-r7) [inline tools/dev/msftidy_docs.rb]: Your logic appears to be inverted.

File: tools/dev/msftidy_docs.rb [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+      warn('Descriptions should be within Vulnerable Application, or an H3 sub-section of Vulnerable Application')
+    end
+
+    unless has_bad_intro
```
Comment (wvu-r7) [inline tools/dev/msftidy_docs.rb]: Your logic appears to be inverted.

File: tools/dev/msftidy_docs.rb [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+  def check_newline_eof
+    if @source !~ /(?:\r\n|\n)\z/m
+      warn('Please add a newline at the end of the file')
+    end
+  end
+
+  # This checks that the H2 headings are in teh right order.
+  def h2_order
+    unless @source =~ /^## Vulnerable Application$.+^## Verification Steps$.+^## Options$.+^## Scenarios$/m
+      warn('H2 headings in incorrect order.  Should be: Vulnerable Application, Verification Steps, Options, Scenarios')
```
Comment (wvu-r7) [inline tools/dev/msftidy_docs.rb]: This is not aware of missing sections.

**Review by wvu — CHANGES_REQUESTED** [review]
At least `chmod +x` and fix the false positives.

File: tools/dev/msftidy_docs.rb (hunk 3 above)
Comment (h00die) [reply tools/dev/msftidy_docs.rb]: :) good catch!

File: tools/dev/msftidy_docs.rb (hunk 5 above)
Comment (h00die) [reply tools/dev/msftidy_docs.rb]: correct.  #6 I asked for @tperry-r7  thoughts on if its OK to have a missing section.  I think Options is most likely the only section that is commonly missing, and I'm personally ok with it as most modules don't have an option that's outside of the 'ordinary'.

Comment (h00die) [reply tools/dev/msftidy_docs.rb]: the other issue is half the comments about things are in PRs the other half in slack,s o also getting lost on where items are getting mentioned or not.

Comment (wvu-r7) [reply tools/dev/msftidy_docs.rb]: I took both warnings into account, so I think it's okay for now. I understood what was intended.

**Review by wvu — APPROVED** [review]
Major blockers fixed.

**Comment by h00die** [comment]
Thanks to @wvu-r7 for some guidance on the long lines in code blocks.  I was thinking some complex regex, he said KISS w/ the flipping Boolean.

**Comment by h00die** [comment]
Found an edge case where code blocks use 4 space, and space at EOL check was hitting on it.  We need to keep those spaces to ensure its one continuous block.

Also started cleaning up more spaces at EOL and other bugs... docs are a mess, it'll be a long road of cleanup.

**Comment by tperry-r7** [comment]
This is great! @h00die !!! Going to merge

**Comment by tperry-r7** [comment]
# Release Notes
This PR adds a new tool: <a href="https://github.com/rapid7/metasploit-framework/blob/master/tools/dev/msftidy_docs.rb">msftidy_docs.rb</a>. It acts like <a href="https://github.com/rapid7/metasploit-framework/blob/master/tools/dev/msftidy.rb">msftidy</a> but is for module documentation.
This is part of a larger effort to standardize the module documentation. See <a href="https://github.com/rapid7/metasploit-framework/pull/12831">#12831<a>. Some of the items it checks for are:
<ul>
<li>Checks to make sure the doc has a corresponding rb/py/go module</li>
<li>Checks ## Vulnerable Application is the first line</li>
<li>Checks to make sure Vulnerable Application, Verification Steps, Scenarios, Options are all present.</li>
</ul>

Thanks to <a href="https://github.com/h00die">@h00die</a>!

**Comment by tperry-r7** [comment]
@h00die 
> Checks the H2s mentioned in 3 are in the right order (@tperry-r7 should it be OK to be missing a section, or all all required? Right now i have all required)


I think we can leave it to look for all. In the future if it starts complaining too much we can allow for sections to be skipped.

File: tools/dev/msftidy_docs.rb (hunk 5 above)
Comment (wvu-r7) [reply tools/dev/msftidy_docs.rb:196]: Fixed in #15140.

**Pull request merged**